README.md
# srt2fcpxml

[![chloe-lam](https://circleci.com/gh/chloe-lam/srt2fcpxml_node.svg?style=shield)](https://circleci.com/gh/chloe-lam/srt2fcpxml_node)

> the srt subtitle file is converted to a final cut pro subtitle

[online document](https://chloe-lam.github.io/srt2fcpxml_node/)

# About FCPXML

[apple official online document](https://developer.apple.com/library/archive/documentation/FinalCutProX/Reference/FinalCutProXXMLFormat/Introduction/Introduction.html)

# installation

```js
npm i srt2fcpxml -g
```

# usage

```bash
Usage: srt2fcpxml [options]

Options:
  -V, --version            output the version number
  -s, --srt <source>       srt file path
  -o, --output <output>    output dir name [default current]
  -r, --rate <rate>        rate: default 30, AKA:23.98,24,25,29.87,30,50,59.94,60
  -e, --event <event>      event name
  -p, --project <project>  project name
  -c, --custom             use custom plugin
  -g, --gap <gap>          gap: number,default 3.6
  -h, --help               display help for command
```

## basic

```bash
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop
```

## FPS rate

```bash
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -r 23.98
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -r 40
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -r 25
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -r 29.87
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -r 30
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -r 50
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -r 59.94
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -r 60
```

## FCPX EventName

```bash
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop -e <eventname>
```

## FCPX projectName

```bash
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop  -p <projectName>
```

## FCPX custom subtitle plugin

```bash
srt2fcpxml -s ~/Desktop/test-vue.srt  -o ~/Desktop  -c
```

![FCPX custom subtitle plugin gif](https://user-images.githubusercontent.com/60084718/100301992-50cb3900-2fd4-11eb-864f-cedbdd2ad5e9.gif)

> using FCPX custom subtitle plugin flag will copy a subtitle plugin to your motion teamplate folder

## thanks:

### go version:

[https://github.com/GanymedeNil/srt2fcpxml](https://github.com/GanymedeNil/srt2fcpxml)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.9](https://github.com/chloe-lam/srt2fcpxml_node/compare/v1.0.8...v1.0.9) (2020-11-26)

### [1.0.8](https://github.com/chloe-lam/srt2fcpxml_node/compare/v1.0.7-0...v1.0.8) (2020-11-26)

### [1.0.7](https://github.com/chloe-lam/srt2fcpxml_node/compare/v1.0.6...v1.0.7) (2020-11-25)

### [1.0.6](https://github.com/chloe-lam/srt2fcpxml_node/compare/v1.0.5...v1.0.6) (2020-11-25)

### [1.0.5](https://github.com/chloe-lam/srt2fcpxml_node/compare/v1.0.4...v1.0.5) (2020-11-25)

### [1.0.4](https://github.com/chloe-lam/srt2fcpxml_node/compare/v1.0.3...v1.0.4) (2020-11-25)

### [1.0.3](https://github.com/chloe-lam/srt2fcpxml_node/compare/v1.0.1...v1.0.3) (2020-11-25)

### 1.0.2 (2020-11-25)


src/parser.ts
import chalk from 'chalk';
import * as fs from 'fs-extra';
import { map, parse } from 'subtitle';
// import { parse, resync, stringify } from 'subtitle';
import { v4 as uuidV4 } from 'uuid';
/* cSpell:disable */
import formater from 'xml-formatter';

/* cSpell:enable */
import { resources } from './constants/resources';
import { CueTemplate, cueType, MainTemplate } from './template';

type ParserParam = {
  readonly srtPath: string;
  readonly outputPath: string;
  readonly eventName: string;
  readonly rateKey?: string;
  readonly projectName?: string;
  readonly gap?: number;
  readonly isUsePlugin?: boolean;
};
function runParse(params: ParserParam) {
  const list = [];
  fs.createReadStream(params.srtPath)
    .pipe(parse())
    .pipe(
      map((node) => {
        if (node.type === 'cue') {
          // eslint-disable-next-line functional/immutable-data
          node.data.text = node.data.text.replace(/"/g, '&#34;');
        }
        return node;
      })
    )
    .on('data', (node) => {
      // eslint-disable-next-line functional/immutable-data
      list.push(node);
    })
    .on('error', (err) => {
      console.log(chalk.blue.bgRed.bold(err));
    })
    .on('finish', () => {
      const lastCue: cueType = list.slice(-1)[0];
      const totalCueTime = lastCue.data.end;
      const fileName = params.srtPath.replace(/(.+)\/(.+)$/, '$2');
      const projectName = params.projectName || fileName;
      /* cSpell:disable */
      const outputPath = params.outputPath + '/' + fileName + '.fcpxml';
      const eventName = params.eventName || 'srt2fcpxml_node';
      /* cSpell:enable */
      const resourceConfig = resources[params.rateKey];
      const gap = params.gap;
      const isUsePlugin = params.isUsePlugin;
      const config = {
        cuesTemplate: CueTemplate(list, resourceConfig, gap, isUsePlugin).join(
          ''
        ),
        project: {
          name: projectName,
          uid: uuidV4(),
          modDate: 'today',
        },
        event: {
          name: eventName,
          uid: uuidV4(),
        },
        totalCueTime: totalCueTime,
        resources: resourceConfig,
        gap,
        isUsePlugin,
      };
      const fcpXMl = MainTemplate(config, resourceConfig);
      const fcpXmlFile = fs.createWriteStream(outputPath);
      /* cSpell:disable */
      const formattedXml = formater(fcpXMl, {
        indentation: '  ',
        filter: (node) => {
          return node;
        },
        collapseContent: true,
        lineSeparator: '\n',
      });
      fcpXmlFile.write(formattedXml);
      fcpXmlFile.end(() => {
        console.log(
          chalk.greenBright.bold(`success! the output file are ${outputPath}`)
        );
      });
    });
}
export const Parser = (params: ParserParam) => {
  return runParse(params);
};


src/template.ts
import { pluginType } from './constants/plugin';
import { resourcesType } from './constants/resources';
export type cueType = {
  readonly type: string;
  readonly data: {
    readonly start: number;
    readonly end: number;
    readonly text: string;
  };
};
type MainTemplateDataType = {
  readonly cuesTemplate: string;
  readonly project: {
    readonly name: string;
    readonly uid: string;
    readonly modDate: string;
  };
  readonly event: {
    readonly name: string;
    readonly uid: string;
  };
  readonly totalCueTime: number;
  readonly resources: resourcesType;
  readonly gap?: number;
  readonly isUsePlugin?: boolean;
};
export const MainTemplate = (
  data: MainTemplateDataType,
  resource: resourcesType
) => {
  const gap = data.gap || 3.6;
  const gapStart =
    gap * resource.frameDurationDenominator * resource.frameDurationMolecular;
  const gapDuration =
    Math.round((data.totalCueTime / 1000) * resource.frameRate) *
    resource.frameDurationMolecular;
  const sequenceDuration =
    Math.round((data.totalCueTime / 1000) * resource.frameRate) *
    resource.frameDurationMolecular;
  const pluginPath = data.isUsePlugin ? pluginType.custom : pluginType.base;
  /* cSpell:disable */
  return `<?xml version="1.0" encoding="UTF-8" ?>
        <!DOCTYPE fcpxml>
        <fcpxml version="1.7">
        <resources>
            <format id="r1" name="FFVideoFormat1920x1080${data.resources.name}" frameDuration="${data.resources.frameDuration}" width="1920" height="1080" colorSpace="1-1-1 (Rec. 709)"></format>
            <effect id="r2" name="字幕" uid="${pluginPath}"></effect>
        </resources>
        <library location="file:///Users/gemer/Movies/test-vue.fcpbundle">
            <event name="${data.event.name}" uid="${data.event.uid}">
                <project name="${data.project.name}" uid="${data.project.uid}" modDate="2020-11-16 13:22:32 +0800">
                    <sequence duration="${sequenceDuration}/${resource.frameDurationDenominator}s" format="r1" tcStart="0s" tcFormat="NDF" audioLayout="stereo" audioRate="48k">
                        <spine>
                            <gap name="空隙" offset="0s" duration="${gapDuration}/${resource.frameDurationDenominator}s" start="${gapStart}/${resource.frameDurationDenominator}s">
                                ${data.cuesTemplate}
                            </gap>
                        </spine>
                    </sequence>
                </project>
            </event>
        </library>
    </fcpxml> 
    `;
};

export const CueTemplate = (
  cue: readonly cueType[],
  resource: resourcesType,
  gap = 3.6,
  isUsePlugin = false
) => {
  return Array.from(cue).map((item, key) => {
    const start = item.data.start / 1000 || 0;
    const end = item.data.end / 1000 || 0;
    const latestGap = gap || 3.6;
    const projectStart =
      latestGap *
      resource.frameDurationDenominator *
      resource.frameDurationMolecular;
    const offset =
      Math.round(start * resource.frameRate) * resource.frameDurationMolecular +
      projectStart;
    const duration =
      (Math.round((end - start) * resource.frameRate) *
        resource.frameDurationMolecular *
        120000.0) /
      resource.frameDurationDenominator;

    return ` <title name="${item.data.text}" lane="1" offset="${offset}/${
      resource.frameDurationDenominator
    }s" ref="r2" duration="${duration}/120000s" start="${projectStart}/${
      resource.frameDurationDenominator
    }s">
                    ${pluginTemplate(isUsePlugin)}
                <text><text-style ref="ts${key + 1}">${
      item.data.text
    }</text-style></text>
                <text-style-def id="ts${key + 1}">
                <text-style font="PingFang SC" fontSize="52" fontFace="Semibold" fontColor="1 1 1 1" bold="1" strokeColor="0.329705 0.329721 0.329713 1" strokeWidth="1" shadowColor="0 0 0 0.75" shadowOffset="3 315" kerning="1.24" alignment="center"/>
                </text-style-def>
             </title>
        `;
  });
};

const pluginTemplate = (isUsePlugin = false): string => {
  const base = ` <param name="位置" key="9999/999166631/999166633/1/100/101" value="0 -450"></param>
  <param name="对齐" key="9999/999166631/999166633/2/354/999169573/401" value="1 (居中)"></param>
  <param name="展平" key="9999/999166631/999166633/2/351" value="1"></param>`;
  const custom = `<param name="位置" key="9999/10199/10201/1/100/101" value="0 -445.000"/>
  <param name="对齐" key="9999/10199/10201/2/354/1002961760/401" value="1 (居中)"/><param name="Scale X" key="9999/1825768416/100/1825768417/2/100" value="0.03333333333333333"/>
  <param name="Scale Y" key="9999/1825768479/100/1825768480/2/100" value="0.03333333333333333"/>
  <param name="Build In" key="9999/10000/2/101" value="0"/>
  <param name="Build Out" key="9999/10000/2/102" value="0"/>
  <param name="Opacity" key="9999/1825768325/10003/10045/1/200/202" value="0.7"/>
  <param name="BG Color" key="9999/1825768325/10003/10045/2/353/113/111" value="0 0 0"/>
  <param name="Out Sequencing" key="9999/10199/10201/4/10233/201/202" value="0 (到)"/>`;
  return isUsePlugin ? custom : base;
};


src/lib/helper.ts
/**
 * checkout file exist
 */
import os from 'os';
import * as path from 'path';

import chalk from 'chalk';
import * as copydir from 'copy-dir';
import * as fs from 'fs-extra';

import { resourcesEnum } from '../constants/resources';

const log = console.log;
export const checkSrtPath = (srtPath: string): boolean => {
  if (!fs.existsSync(srtPath)) {
    log(chalk.red("the srt file doesn't exist at that path"));
    return false;
  }
  return true;
};

export const checkRate = (rate: string): boolean => {
  const keys = Object.keys(resourcesEnum);
  if (!keys.includes(rate)) {
    log(
      chalk.red(
        "the input rate doesn't match, example: 23.98,24,25,29.87,30,50,59.94,60"
      )
    );
    return false;
  }
  return true;
};
export const formateRateKey = (rate: string): string => {
  if (rate) {
    return 'p' + (parseFloat(rate) * 100).toString();
  } else {
    return resourcesEnum.p3000;
  }
};

export const copyPlugin = () => {
  const titlesPath = `${os.homedir()}/Movies/Motion Templates.localized/Titles.localized`;
  const pluginPath = path.resolve(
    __dirname,
    '../../../plugin/adaptive-background'
  );
  if (!fs.existsSync(titlesPath)) {
    /* cSpell:disable */
    log(
      chalk.red(
        'hey, your FCPX Motion is not a default path, please copy by yourself'
      )
    );
    /* cSpell:enable */
  }
  copydir.sync(
    pluginPath,
    titlesPath + '/adaptive-background',
    {
      utimes: true, // keep add time and modify time
      mode: true, // keep file mode
      cover: true, // cover file when exists, default is true
    },
    (error) => {
      if (error)
        log(chalk.red('hey, somethings wrong,please copy plugin by yourself'));
    }
  );
  log(chalk.greenBright('copy custom plugin done!'));
  return null;
};


src/constants/resources.ts
export enum resourcesEnum {
  p2398 = 'p2398',
  p2400 = 'p2400',
  p2500 = 'p2500',
  p2997 = 'p2997',
  p3000 = 'p3000',
  p5000 = 'p5000',
  p5994 = 'p5994',
  p6000 = 'p6000',
}
export type resourcesType = {
  readonly name: string;
  readonly duration: string;
  readonly frameDuration: string;
  readonly frameRate: number;
  readonly frameDurationMolecular: number;
  readonly frameDurationDenominator: number;
};
export const resources = {
  p2398: {
    name: 'p2398',
    duration: '2400s',
    frameDuration: '1001/24000s',
    frameRate: 23.98,
    frameDurationMolecular: 1001,
    frameDurationDenominator: 24000,
  },
  p2400: {
    name: 'p2400',
    duration: '2400s',
    frameDuration: '100/2400s',
    frameRate: 24,
    frameDurationMolecular: 100,
    frameDurationDenominator: 2400,
  },
  p2500: {
    name: 'p2500',
    duration: '2500s',
    frameDuration: '100/2500',
    frameRate: 25,
    frameDurationMolecular: 100,
    frameDurationDenominator: 2500,
  },
  p2997: {
    name: 'p2997',
    duration: '3000s',
    frameDuration: '1001/30000s',
    frameRate: 29.97,
    frameDurationMolecular: 1001,
    frameDurationDenominator: 30000,
  },
  p3000: {
    name: 'p3000',
    duration: '3000s',
    frameDuration: '100/3000s',
    frameRate: 30,
    frameDurationMolecular: 100,
    frameDurationDenominator: 3000,
  },
  p5000: {
    name: 'p5000',
    duration: '5000s',
    frameDuration: '100/5000s',
    frameRate: 50,
    frameDurationMolecular: 100,
    frameDurationDenominator: 5000,
  },
  p5994: {
    name: 'p2398',
    duration: '6000s',
    frameDuration: '1001/60000s',
    frameRate: 50,
    frameDurationMolecular: 1001,
    frameDurationDenominator: 60000,
  },
  p6000: {
    name: 'p6000',
    duration: '6000s',
    frameDuration: '100/6000s',
    frameRate: 60,
    frameDurationMolecular: 100,
    frameDurationDenominator: 6000,
  },
};


src/index.ts
#!/usr/bin/env node
export * from './lib/async';
export * from './lib/number';
export { Parser } from './parser';
import * as path from 'path';

import confirm from '@inquirer/confirm';
import * as root from 'app-root-path';
import chalk from 'chalk';
import { program } from 'commander';
import * as fs from 'fs-extra';

import {
  checkRate,
  checkSrtPath,
  copyPlugin,
  formateRateKey,
} from './lib/helper';
import { Parser } from './parser';

const packageJson = fs.readJsonSync(
  path.resolve(__dirname, '../../', 'package.json')
);
program
  .version(packageJson.version)
  .requiredOption('-s, --srt <source>', 'srt file path')
  .option('-o, --output <output>', 'output dir name [default current]')
  .option(
    '-r, --rate <rate>',
    'rate: default 30, AKA:23.98,24,25,29.87,30,50,59.94,60'
  )
  .option('-e, --event <event>', 'event name')
  .option('-p, --project <project>', 'project name')
  .option('-c, --custom', 'use custom plugin')
  .option('-g, --gap <gap>', 'gap: number,default 3.6');

program.parse(process.argv);

const mainProcess = (isUsePlugin = false) => {
  if (program.opts().srt) {
    const srtParam = program.opts().srt;
    const outputParam = program.opts().output || root.path;
    const rateParam = formateRateKey(program.opts().rate);
    /* cSpell:disable */
    const eventName = program.opts().event || 'srt2fcpxml_node';
    const projectName = program.opts().project;
    const gap = parseFloat(program.opts().gap);

    if (checkSrtPath(srtParam) && checkRate(rateParam)) {
      Parser({
        srtPath: srtParam,
        outputPath: outputParam,
        eventName,
        rateKey: rateParam,
        projectName,
        gap,
        isUsePlugin,
      });
    }
  }
};
if (program.opts().custom) {
  (async () => {
    const answer = await confirm({
      message:
        'Hey, Do you want use custom subtitle fcpx plugin? \n will copy file to your \n ~/Movies/Motion Templates.localized/Titles.localized ',
    });
    if (answer) {
      copyPlugin();
    } else {
      console.log(
        chalk.yellowBright.bold('please copy that plugin by yourself')
      );
    }
    mainProcess(true);
  })();
} else {
  mainProcess();
}
